File: fortran/backup/lpopul.f90
!
!
 SUBROUTINE LPopul(n,L,D,Nb,vpl,va,xl)
! **********************************************************************
! *                                                                    *
! * Initialize geometry, velocities and parameters for MD simulation   *
! *                                                                    *
! **********************************************************************
! *                                                                    *
! * Author:     Boris Fackovec                                         *
! * Date:       30/01/2016                                             *
! * Version:    1.0                                                    *
! *                                                                    *
! **********************************************************************
!
   IMPLICIT NONE
   INTEGER, INTENT(IN) :: n
   DOUBLE PRECISION, INTENT(IN), DIMENSION(n,n) :: L
   DOUBLE PRECISION, INTENT(IN), DIMENSION(n,n) :: D
   LOGICAL, INTENT(IN), DIMENSION(n,n) :: Nb
   DOUBLE PRECISION, INTENT(IN), DIMENSION(n) :: vpl
   LOGICAL, INTENT(IN), DIMENSION(n) :: va
   DOUBLE PRECISION, INTENT(OUT) :: x
!
! -------------------------------------------------------------------
!
   INTEGER :: na, i, j, k
   DOUBLE PRECISION :: pl, pl1, pl2, pltemp
   INTEGER, DIMENSION(n) :: nind
   DOUBLE PRECISION, DIMENSION(n) :: vpla, vplnow, vpltemp
   DOUBLE PRECISION, DIMENSION(n**2) :: vf1
   LOGICAL :: bf1
   LOGICAL, DIMENSION(n**2) :: vbf1
!
! -------------------------------------------------------------------
!
!
!  find the size of state A and calculate its population
!
   na = 0
   cycCountA: DO i=1,n
      IF (va(i)) THEN 
         na = na + 1 
         vpla(na) = vpl(i)
         vind(na) = i
      ENDIF
   ENDDO cycCountA
   CALL LogSumExp(na,vpl,pl1)
   CALL LogDiffExp(0.0d0,pl1,pl2)
!
!  get the population from log transition matrix
!
   cycSumCols: DO i=1,na
      cycSumRows: DO j=1,na
         vpltemp(j) = L(vind(i),vind(j))+vpl(vind(j))-pl1
      ENDDO cycSumRows
      CALL LogSumExp(na,vpltemp,pltemp)
      vplnow(i) = pltemp
   ENDDO cycSumCols
   CALL LogSumExp(na,vplnow,pl)
   xl = logDiffExp(pl,pl1)-pl2
   IF (xl.LT.-3d0) THEN
      RETURN
   ENDIF
!
!   get the population from log(1-t)
!
   cycAddCols: DO i=1,na
      cycAddRows: DO j=1,na
         k = (i-1)*na+j
         vf1(k) = vpl(vind(i)) + vpl(vind(j)) + D(vind(i),vind(j))
         vb1f(k) = N(vind(i),vind(j))
      ENDDO cycAddRows
   ENDDO cycAddCols
   CALL LogSumDiff(na**2,vf1,vbf1,pl,bf1)
   CALL LogDiffExp(0.0d0,pl,xl)
   xl = xl - pl1 - pl2
!
   RETURN
!
! -------------------------------------------------------------------
!
 END SUBROUTINE LPopul
!

File: new/lpopul.m
function x = lpopul(L,D,N,vpl,va)
% calculates numerator of progress of the relaxation (current population in state A - the equilibrium population)
% first try simply Tl
vi = find(va);
n = size(vi,1);
vpl = vpl(vi); L = L(vi,vi); D = D(vi,vi); N = N(vi,vi);
pl1 = logSumExp(vpl); pl2 = subFrom1(pl1);
vplnow = zeros(n,1);
for i=1:n
  vplnow(i) = logSumExp(L(i,:)'+vpl-pl1);
endfor
pl = logSumExp(vplnow);
x = logDiffExp(pl,pl1)-pl2
if (x>-3)
  return;
endif

From1 = zeros(n^2,2);
for i=1:n
  for j=1:n
    k = (i-1)*n+j;
    From1(k,:) = [vpl(i) + vpl(j) + D(i,j), N(i,j)];
  endfor
endfor
vojoj = logSumDiff(From1);
x = vojoj(1)-pl1-pl2
endfunction
